Path search with cost function

mgttlinger opened the pull request:
Functionality for finding the cheapest path between two nodes according to a cost function that sees both nodes and edge labels.

mgttlinger committed:
Added tests to back the claims in the opened issues

mgttlinger committed:
Merge branch 'master' into tests

mgttlinger committed:
Path search with costFkt

mgttlinger committed:
Cherry-picked generators for testing

mgttlinger committed:
Refactored for readability

mgttlinger committed:
Added test for new functionality

mgttlinger committed:
Fixed type of LPath to have the right number of nodes and labels.

mgttlinger committed:
Merge branch 'lpath' into tests

mgttlinger committed:
Removed unnecessary Option

mgttlinger committed:
Merge branch 'lpath' into tests

mgttlinger committed:
Made tests compile with the new LPath type

rossabaker commented on core/src/main/scala/Graph.scala:
```diff
@@ -903,6 +903,30 @@ case class Graph[N,A,B](rep: GraphRep[N,A,B]) {
+   * Cheapest path from vertex `s` to vertex `t` under the cost function `costFkt` with labels
+   * @group bfs   
+   */
+  def cheapestPath[C](s: N, t: N, costFkt: (LNode[N,A],B,LNode[N,A]) => C)(implicit num: Numeric[C]): Option[LPath[N,B]] = {
```
This could be a little more generic as `C: Monoid: Order`.  I imagine `C` would _usually_ be numeric in the real world, but it doesn't have to be.

rossabaker commented on core/src/main/scala/Graph.scala:
```diff
@@ -903,6 +903,30 @@ case class Graph[N,A,B](rep: GraphRep[N,A,B]) {
+   */
+  def cheapestPath[C](s: N, t: N, costFkt: (LNode[N,A],B,LNode[N,A]) => C)(implicit num: Numeric[C]): Option[LPath[N,B]] = {
+    require(contains(s))
+    require(contains(t))
```
Could we return `None` here, or is it important to distinguish a missing vertex from an empty searchtree?  The exceptions make me a bit uncomfortable.

timperrett replied:
Yeah, ideally we don't throw exceptions but instead codify things into the type system :-)

mgttlinger replied:
Rewrite all of this in idris and I will happily encode that in the signature :)

rossabaker replied:
Well, until someone publishes quividris, my question remains. :)  Would a `None` be an acceptable result, or do we need a third value to return?

mgttlinger replied:
Sorry for not answering this sooner. I think that would be perfectly fine.

rossabaker reviewed (commented):
Thanks.  This seems mostly reasonable to me.  I raised a couple questions, but I'm also not the most experienced user of this library, so I'm open to being called a fool. :)

adelbertc committed:
Fix GDecomp#redecorate

- Redecoration includes the focus of the decomposition when acting on
the rest of the graph
- Fix GDecomp generator so incorrect instances cannot be built (this
should probably be a smart constructor)

Commit:
Fixed comments

Commit:
Fixed labeled search

rossabaker reviewed (approved):
:+1: Thanks

rossabaker commented:
Uh-oh.  This doesn't compile together with the changes to `LPath` in #23.  Is one more urgent to you than the other, or do you want to fix them up so they can both go?

#23 and #26 are breaking changes.  I'm planning to open a quiver6 branch so we only bump the major version once with our automated release process.  We can target that with whatever comes of your open PRs and get a release cut before the cats port.

rossabaker committed:
Apply @joroKr21's fix to genGDecomp

rossabaker committed:
Merge remote-tracking branch 'adelbertc/fix-redecorate' into quiver-6

mgttlinger commented:
I can fix this tomorrow. Ideally the LPath stuff would be merged by then. If not, I'll cherry pick it into this branch as well.

rossabaker commented:
Great.  I'll merge #23 into the quiver-6 branch now, and you can target that branch with your fixed up version of this, and we'll be ready to release.

rossabaker committed:
Merge remote-tracking branch 'mgttlinger/lpath' into quiver-6

Commit:
Merge branch 'quiver-6' into path-with-costfkt

Commit:
Changed path search with cost function to the noe LPath type

mgttlinger commented:
Closing because now targeting quiver-6 in #28
